File: code/Lambda.py
#!/usr/bin/env python3

from types import SimpleNamespace

class Term(SimpleNamespace):
  pass

class Var(Term):
  def __init__(self, vname):
    self.name = vname

  def eval(self, context):
    return self.substitute(context)

  def substitute(self, context):
    return context[self.name] if self.name in context else self

class Abs(Term):
  def __init__(self, vname, term):
    self.vname = vname
    self.term = term

  def eval(self, context):
    return self.substitute(context)

  def substitute(self, context):
    old = None
    if self.vname in context:
      old = context[self.vname]
      del context[self.vname]
    result = Abs(self.vname, self.term.substitute(context))
    if old is not None:
      context[self.vname] = old
    return result

class App(Term):
  def __init__(self, term1, term2):
    self.term1 = term1
    self.term2 = term2

  def eval(self, context):
    term1 = self.term1.eval(context)
    term2 = self.term2.eval(context)
    if type(term1) == Abs:
      old = None
      if term1.vname in context:
        old = context[term1.vname]
      context[term1.vname] = term2
      result = term1.term.eval(context)
      if old is not None:
        context[term1.vname] = old
      else:
        del context[term1.vname]
      return result
    else:
      return App(term1, term2)

  def substitute(self, context):
    return App(self.term1.substitute(context), self.term2.substitute(context))

def main():
  tests = \
      [ App(Abs('x', App(Var('x'), Var('x'))), Var('y'))
      , App(App(Abs('x',Abs('y',Var('x'))), Var('a')), Var('b'))
      , App(App(Abs('x',Abs('y',Var('y'))), Var('a')), Var('b')) ]
  for t in tests:
    print('before',t)
    print('after',t.eval({}))
    print('---')

if __name__ == '__main__':
  main()

File: code/Lambda.java
interface Env {
  Term get(String id);
  void del(String id);
  void add(String id, Term t);
}

interface Term {
  Term eval(Env env);
  Term subst(Env env);
}

class Var implements Term {
  String id;
  Var(String id) {this.id = id;}
  @Override public Term eval(Env env) { return subst(env); }
  @Override public Term subst(Env env) {
    Term r = env.get(id);
    return r == null? this : r;
  }
}

class Abs implements Term {
  String vname;
  Term body;
  Abs(String vname, Term body) {
    this.vname = vname;
    this.body = body;
  }
  @Override public Term eval(Env env) { return subst(env); }
  @Override public Term subst(Env env) {
    Term old = env.get(vname);
    if (old != null) env.del(vname);
    Term result = new Abs(vname, body.subst(env));
    if (old != null) env.add(vname, old);
    return result;
  }
}

class App implements Term {
  Term t1, t2;
  App(Term t1, Term t2) { this.t1 = t1; this.t2 = t2; }
  @Override public Term eval(Env env) {
    Term t1e = t1.eval(env);
    Term t2e = t2.eval(env);
    if (t1e instanceof Abs) {
      Term old = env.get(t1e.vname);
      //if (old != null) env.del(t1e.vname);
      env.add(t1e.vname, t2e);
      Term result = t1e.body.eval(env);
      if (old != null) {
        env.add(t1e.vname, old);
      } else {
        env.del(t1e.vname);
      }
      return result;
    } else {
      return new App(t1e, t2e);
    }
  }

  @Override public Term subst(Env env) {
    return null;
  }
}

public class Lambda {
}
